### PR TITLE
[SPARK-37397][PYTHON] Inline annotations for pyspark.ml.base

### PR DESCRIPTION
**Pull request by zero323**
### What changes were proposed in this pull request?

Migration of type annotation for `pyspark.ml.base` from stub file to inline hints.

### Why are the changes needed?

As a part of ongoing type hints migration.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests + new data tests.

### CODE REVIEW

File: python/pyspark/ml/classification.pyi [hunk 1]
```diff
@@ -820,6 +820,7 @@ class OneVsRest(
         weightCol: Optional[str] = ...,
         parallelism: int = ...,
     ) -> OneVsRest: ...
+    def _fit(self, dataset: DataFrame) -> OneVsRestModel: ...
```
Comment (zero323) [inline python/pyspark/ml/classification.pyi:823]: This is added to suppress abstract method warnings.

File: python/pyspark/ml/classification.pyi [hunk 2]
```diff
@@ -832,6 +833,7 @@ class OneVsRest(
 class OneVsRestModel(Model, _OneVsRestParams, MLReadable[OneVsRestModel], MLWritable):
     models: List[Transformer]
     def __init__(self, models: List[Transformer]) -> None: ...
+    def _transform(self, dataset: DataFrame) -> DataFrame: ...
```
Comment (zero323) [inline python/pyspark/ml/classification.pyi:836]: Ditto.

File: python/pyspark/ml/param/__init__.py [hunk 3]
```diff
@@ -492,7 +493,7 @@ def _dummy() -> "Params":
         dummy.uid = "undefined"
         return dummy
 
-    def _set(self, **kwargs: Any) -> "Params":
+    def _set(self: P, **kwargs: Any) -> P:
```
Comment (zero323) [inline python/pyspark/ml/param/__init__.py:496]: This is allow us to hint specific subclass of `Params`, that depends on the input.

File: python/pyspark/ml/param/__init__.py [hunk 4]
```diff
@@ -529,7 +530,7 @@ def _setDefault(self, **kwargs: Any) -> "Params":
             self._defaultParamMap[p] = value
         return self
 
-    def _copyValues(self, to: "Params", extra: Optional["ParamMap"] = None) -> "Params":
+    def _copyValues(self, to: P, extra: Optional["ParamMap"] = None) -> P:
```
Comment (zero323) [inline python/pyspark/ml/param/__init__.py:533]: Ditto.

File: python/pyspark/ml/param/__init__.py [hunk 5]
```diff
@@ -513,7 +514,7 @@ def clear(self, param: Param) -> None:
         if self.isSet(param):
             del self._paramMap[param]
 
-    def _setDefault(self, **kwargs: Any) -> "Params":
+    def _setDefault(self: P, **kwargs: Any) -> P:
```
Comment (zero323) [inline python/pyspark/ml/param/__init__.py:517]: Ditto.

File: python/pyspark/ml/pipeline.pyi [hunk 6]
```diff
@@ -33,10 +33,12 @@ from pyspark.ml.util import (  # noqa: F401
     MLWritable as MLWritable,
     MLWriter as MLWriter,
 )
+from pyspark.sql.dataframe import DataFrame
 
 class Pipeline(Estimator[PipelineModel], MLReadable[Pipeline], MLWritable):
     stages: List[PipelineStage]
     def __init__(self, *, stages: Optional[List[PipelineStage]] = ...) -> None: ...
+    def _fit(self, dataset: DataFrame) -> PipelineModel: ...
```
Comment (zero323) [inline python/pyspark/ml/pipeline.pyi:41]: To avoid errors on missing abstract implementation.

**Comment by zero323** [comment]
cc @itholic @ueshin @xinrong-databricks FYI.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
cc @WeichenXu123 FYI if you find some time to review.

**Review by itholic — APPROVED** [review]
Thanks!

**Comment by zero323** [comment]
I'll wait for few more days, to give other mentioned folks to provide feedback and, if there is no response, I'll go ahead and merge it to be able to proceed with the rest of `pyspark.ml`.

**Comment by dongjoon-hyun** [comment]
Thank you all. Merged to master. Please proceed to the rest of `pyspark.ml`, @zero323 .

**Comment by zero323** [comment]
Thank you @dongjoon-hyun!

**Pull request closed without merging**